tests/testthat/test-super.R

```diff
@@ -0,0 +1,39 @@
+
+describe("super()", {
+  it("overrides dispatch, matching exact class", {
+    foo1 <- new_class("foo1")
+    foo2 <- new_class("foo2", foo1)
+    foo3 <- new_class("foo3", foo2)
+
+    bar <- new_generic("bar", "x")
+    method(bar, foo1) <- function(x) 1
+    method(bar, foo3) <- function(x) 3
+
+    expect_error(bar(super(foo3())), "Can't find method")
```

Comment:
I feel like this is inconsistent

``` r
foo1 <- new_class("foo1")
foo2 <- new_class("foo2", foo1)
foo3 <- new_class("foo3", foo2)

bar <- new_generic("bar", "x")
method(bar, foo1) <- function(x) 1
method(bar, foo3) <- function(x) 3

# this doesnt work
bar(super(foo3()))
#> Error: Can't find method for generic `bar()` with dispatch classes:
#> - x: foo2

# but this does
bar(foo2())
#> [1] 1
```

Reply:
Do you think we could talk about this one in the next OOP working group meeting?

To me it feels like `super(foo3, foo2)` says "I return an object of class foo2", meaning that `bar(foo2())` is identical to `bar(super(foo3(), foo2))` in my head

I am interpreting your argument as: If the user wanted the `foo1` behavior, then they should have done `super(foo3, foo1)`, which also feels reasonable (and more explicit). But it still might be nice to discuss this point.

Reply:
BTW this is consistent with `convert()` if you do `convert(x, to = foo2)`, you don't expect to get an instance of `foo1` back.

Reply:
But you do get something back that you can call `bar()` on, and that will dispatch to the `foo1` method?

```r
x <- foo3()

bar(x)
#> [1] 3

bar(convert(x, to = foo2))
#> [1] 1

bar(super(x, to = foo2))
#> Error: Can't find method for generic `bar()` with dispatch classes:
#> - x: foo2
```

Reply:
Right. It's not directly analogous, but I think it's in the same spirit.

Reply:
Ok, let's change this so that `super(..., to = foo2)` behaves exactly like a `foo2`